src/hotspot/os/aix/os_aix.cpp

```diff
@@ -3033,29 +3049,134 @@ void os::jfr_report_memory_info() {}
+    return -1;
+
+  char *path2 = strdup (path);
+  // if exist, strip off trailing (shr_64.o) or similar
+  int idx = strlen(path2) - 1;
+  if (path2[idx] == ')') {
+    while (path2[idx] != '(' && idx > 0) idx--;
+    if (idx > 0)
+      path2[idx] = 0;
+  }
+
+  int ret = -1;
+  if (path2[0] == '/' ||
+      (path2[0] == '.' && (path2[1] == '/' ||
+                          (path2[1] == '.' && path2[2] == '/')))) {
+    ret = stat64x(path2, stat);
+    free (path2);
+    return ret;
   }
 
   const char* env = getenv("LIBPATH");
-  if (env == nullptr || *env == 0)
+  if (env == nullptr || *env == 0) {
+    free (path2);
     return -1;
+  }
 
-  int ret = -1;
   size_t libpathlen = strlen(env);
   char* libpath = NEW_C_HEAP_ARRAY(char, libpathlen + 1, mtServiceability);
-  char* combined = NEW_C_HEAP_ARRAY(char, libpathlen + strlen(path) + 1, mtServiceability);
+  char* combined = NEW_C_HEAP_ARRAY(char, libpathlen + strlen(path2) + 1, mtServiceability);
   char *saveptr, *token;
   strcpy(libpath, env);
   for (token = strtok_r(libpath, ":", &saveptr); token != nullptr; token = strtok_r(nullptr, ":", &saveptr)) {
-    sprintf(combined, "%s/%s", token, path);
+    sprintf(combined, "%s/%s", token, path2);
     if (0 == (ret = stat64x(combined, stat)))
       break;
   }
 
   FREE_C_HEAP_ARRAY(char*, combined);
   FREE_C_HEAP_ARRAY(char*, libpath);
+  free (path2);
   return ret;
 }
+
+// specific AIX versions for ::dlopen() and ::dlclose(), which handles the struct g_handletable
+// filled by os::dll_load(). This way we mimic dl handle equality for a library
+// opened a second time, as it is implemented on other platforms.
+void* os::Aix::dlopen(const char* filename, int Flags) {
+  void* result;
+  struct stat64x libstat;
+
+  if (os::Aix::stat64x_via_LIBPATH(filename, &libstat)) {
+    // file with filename does not exist
+    result = ::dlopen(filename, Flags);
+    if (result != nullptr) {
+      assert(false, "dll_load: Could not stat() file %s, but dlopen() worked; Have to improve stat()", filename);
+    }
+  }
+  else {
+    int i = 0;
+    pthread_mutex_lock(&g_handletable_mutex);
+    // check if library belonging to filename is already loaded.
+    // If yes use stored handle from previous ::dlopen() and increase refcount
+    for (i = 0; i < g_handletable_used; i++) {
+      if (g_handletable[i].handle &&
+          g_handletable[i].inode == libstat.st_ino &&
+          g_handletable[i].devid == libstat.st_dev) {
+        g_handletable[i].refcount++;
+        result = g_handletable[i].handle;
+        break;
+      }
+    }
+    if (i == g_handletable_used) {
+      // library not still loaded. Check if there is space left in array
+      // to store new ::dlopen() handle
+      if (g_handletable_used == max_handletable) {
+        // Array is already full. No place for new handle. Cry and give up.
+        pthread_mutex_unlock(&g_handletable_mutex);
+        assert(false, "max_handletable reached");
+        return nullptr;
```

Comment:
Please, for the sake of release code, hand in an error buffer and fill it with something that makes sense, eg. "too many libraries loaded".
The assert is still okay, I guess, since we don't expect it to fire during tests; if it does fire, it may indicate a problem in our handle table logic or a wrong assumption about handle équality.